lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -0,0 +1,254 @@
+    {
+        self.autodiscovery_provider = Some(Arc::new(provider));
+        self
+    }
+}
+
+#[async_trait]
+impl SourceBuilder for ChecksConfiguration {
+    async fn build(&self, _context: ComponentContext) -> Result<Box<dyn Source + Send>, GenericError> {
+        match &self.autodiscovery_provider {
+            Some(autodiscovery) => Ok(Box::new(ChecksSource {
+                autodiscovery: Arc::clone(autodiscovery),
+                check_runners: self.check_runners,
+            })),
+            None => Err(generic_error!("No autodiscovery provider configured.")),
+        }
+    }
+
+    fn outputs(&self) -> &[OutputDefinition] {
+        static OUTPUTS: [OutputDefinition; 1] = [OutputDefinition::default_output(DataType::Metric)];
+
+        &OUTPUTS
+    }
+}
+
+impl MemoryBounds for ChecksConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        builder.minimum().with_single_value::<ChecksSource>("component struct");
+    }
+}
+
+struct ChecksSource {
+    autodiscovery: Arc<dyn AutodiscoveryProvider + Send + Sync>,
+    check_runners: usize,
+}
+
+trait CheckBuilder {
+    fn build_check(&self, check_id: &str, check_request: &Config) -> Option<Arc<dyn Check + Send + Sync>>;
+}
+
+/// A no-op implementation of CheckBuilder
+struct NoOpCheckBuilder;
+
+impl CheckBuilder for NoOpCheckBuilder {
+    fn build_check(&self, _check_id: &str, _check_request: &Config) -> Option<Arc<dyn Check + Send + Sync>> {
+        None
+    }
+}
+
+struct ChecksCollector {
+    shutdown_handle: DynamicShutdownHandle,
+    health: Health,
+    event_rx: Receiver<AutodiscoveryEvent>,
+    check_builders: Vec<Arc<dyn CheckBuilder + Send + Sync>>,
+    checks: HashSet<String>,
+    scheduler: Scheduler,
+}
+
+impl ChecksCollector {
+    pub fn new(
+        shutdown_handle: DynamicShutdownHandle, health: Health, event_rx: Receiver<AutodiscoveryEvent>,
+        check_builders: Vec<Arc<dyn CheckBuilder + Send + Sync>>, check_runners: usize,
+    ) -> Self {
+        Self {
+            shutdown_handle,
+            health,
+            event_rx,
+            check_builders,
+            checks: HashSet::new(),
+            scheduler: Scheduler::new(check_runners),
+        }
+    }
+
+    pub async fn run(mut self) {
+        let check_dispatcher_shutdown_coordinator = DynamicShutdownCoordinator::default();
+
+        loop {
+            select! {
+                _ = &mut self.shutdown_handle => {
+                    debug!("Received shutdown signal.");
+                    self.scheduler.shutdown().await;
+                    break
+                },
+
+                _ = self.health.live() => continue,
+
+                event = self.event_rx.recv() => match event {
+                        Ok(event) => {
+                            match event {
+                                AutodiscoveryEvent::Schedule { config } => {
+                                    let digest = &config.digest();
```

Comment:
I would drop taking the reference here (so make it just `config.digest()`) and then take the reference at the point of usage down when you're calling `config.id(...)`.